File: backend/Models/Map.cs
namespace snake_game.Models;

public class Map
{
    public int Id { get; set; }
    public string Name { get; set; }
    public double Rating { get; set; } = 0;
    public bool IsPublished { get; set; }

    public List<MapRating> MapRatings { get; set; }
    public User Creator { get; set; }
    public List<MapObstacle> MapObstacles { get; set; }
}

public record MapDto(
    int Id,
    string Name,
    List<MapObstacleDto> MapObstacles,
    double Rating,
    UserDto Creator
);

public record CreateMapDto(string Name, List<CreateMapObstacleDto> MapObstacles);

public record UpdateMapDto(int MapRating, int UserId);

public static class MapMapper
{
    public static MapDto ToDto(this Map map) =>
        new MapDto(
            map.Id,
            map.Name,
            map.MapObstacles.Select(o => o.ToDto()).ToList(),
            map.Rating,
            map.Creator.ToDto()
        );

    public static Map FromCreateDto(
        this CreateMapDto dto,
        List<MapObstacle> mapObstacles,
        User creator
    )
    {
        return new Map
        {
            Name = dto.Name,
            MapObstacles = mapObstacles,
            IsPublished = false,
            Creator = creator,
        };
    }
}

File: SnakeGame.Client/src/shared/interfaces/map/Map.ts
import { UserDto } from "..";

export interface PointDto {
  id: number;
  X: number;
  Y: number;
}

export interface ObstacleDto {
  id: number;
  points: PointDto[];
}

export interface MapObstacleDto {
  mapObstacleId: number;
  position: PointDto;
  obstacleId: number;
}

export interface MapDto {
  id: number;
  name: string;
  rating: number;
  mapObstacles: MapObstacleDto[];
  isPublished: boolean;
  creator: UserDto;
}
